### ISSUE TITLE
failed to find rejected notification

### ISSUE THREAD

**Issue by killme2008** [relayrides/pushy#149]
Found a new error log in our system:
```
failed to find rejected notification with sequence number 0 (buffer has range null to null);
 this may mean the sent notification buffer is too small. Please report this as a bug.
```

**Comment by varigamy** [comment]
A set of similar errors occur while sending iOS notifications on a DEVELOPMENT certificate:
connection-2460 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.

**Comment by jchambers** [comment]
Do you have a sense of *when* this is happening? My immediate guess is that the APNs gateway is sending a "shutting down for maintenance" response even though you haven't sent any notifications yet. In any case, we can make that error message less null-y if the buffer is empty.

**Comment by jchambers** [comment]
On a related note, do you two have timestamps for these log entries? If they're at the same time, that'd be interesting to know.

**Comment by varigamy** [comment]
I can show you a part of the log file: we don't display milliseconds in the time format, only hours:minutes:seconds; likely it can be enough for you to know. Errors are not stopping. 

And one more important thing is that when 'errors are arriving' and at that moment we send a push notification (add SimpleApnsPushNotification object to the PushManagers' queue), our Apple device receives multiple notifications (every time can be different number).

ERROR [2015-02-11 08:16:47] nioEventLoopGroup-11-2,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: 35#maldiVES!-connection-58065 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:47] nioEventLoopGroup-13-1,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: pratchett-connection-27386 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:47] nioEventLoopGroup-10-1,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: androidsucks-connection-153906 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:47] nioEventLoopGroup-13-2,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: pratchett-connection-27387 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:47] nioEventLoopGroup-11-1,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: 35#maldiVES!-connection-58066 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:48] nioEventLoopGroup-10-2,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: androidsucks-connection-153907 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:53] nioEventLoopGroup-13-1,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: pratchett-connection-27388 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:53] nioEventLoopGroup-11-2,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: 35#maldiVES!-connection-58067 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:53] nioEventLoopGroup-10-1,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: androidsucks-connection-153908 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:53] nioEventLoopGroup-13-2,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: pratchett-connection-27389 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:53] nioEventLoopGroup-11-1,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: 35#maldiVES!-connection-58068 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:53] nioEventLoopGroup-10-2,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: androidsucks-connection-153909 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:58] nioEventLoopGroup-13-1,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: pratchett-connection-27390 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:58] nioEventLoopGroup-11-2,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: 35#maldiVES!-connection-58069 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:58] nioEventLoopGroup-10-1,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: androidsucks-connection-153910 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:58] nioEventLoopGroup-13-2,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: pratchett-connection-27391 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.
ERROR [2015-02-11 08:16:58] nioEventLoopGroup-11-1,com.relayrides.pushy.apns.ApnsConnection - com.bitplaces.ntf.notification.service.sender.IOsNotificationSender.Certificate ID: 35#maldiVES!-connection-58070 failed to find rejected notification with sequence number 0 (buffer has range null to null); this may mean the sent notification buffer is too small. Please report this as a bug.

Thank you for your feedback

**Comment by jchambers** [comment]
Marked #150 as a duplicate of this. It does sound like a serious issue, and we'll do our best to get to the bottom of it as soon as possible.

**Comment by jchambers** [comment]
@killme2008 @varigamy @U-238 @dwxw: Are you folks seeing this in production APNs environments, or only in sandbox environments? As near as I can tell (and this may just be luck so far), Apple has changed the behavior of their sandbox servers to *not* include the sequence number rejected notifications when the token is missing (which is how we "poison" connections to close them). Other rejections seem to be unchanged.

Of course, we should probably try to handle this case more gracefully in the event that Apple rolls this change out to their production environment. Still, I'll file a radar if we all agree that's what seems to be happening.

@matzew ^ heads up on this, by the way. Do you folks do the send-known-bad-notification-to-close-connections thing, too, or is that just us?

**Comment by dwxw** [comment]
I've only seen it on sandbox.  Our live stuff seems to be fine although it doesn't happen every time on sandbox either.

**Comment by jchambers** [comment]
Okay. I've filed http://openradar.appspot.com/radar?id=5843845044502528.

**Comment by kenmaru** [comment]
I'm getting this error too. 
I have a test program that uses sandbox which sends one push notification.
The program fails to stop and it keeps sending the same push notification over and over.

2015-02-18 09:18:29,910 [main] INFO  com.relayrides.pushy.apns.PushManager - ExamplePushManager shutting down.
2015-02-18 09:18:29,911 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-0 sending known-bad notification to shut down.
2015-02-18 09:18:29,974 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - APNs gateway rejected notification with sequence number 0 from ExamplePushManager-connection-0 (MISSING_TOKEN).
2015-02-18 09:18:29,975 [nioEventLoopGroup-2-1] ERROR com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-0 failed to find rejected notification with sequence number 0 (buffer has range 1 to 1); this may mean the sent notification buffer is too small. Please report this as a bug.
2015-02-18 09:18:29,981 [pool-1-thread-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-1 beginning connection process.
2015-02-18 09:18:30,044 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-1 connected; waiting for TLS handshake.
2015-02-18 09:18:30,227 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-1 successfully completed TLS handshake.
2015-02-18 09:18:30,231 [nioEventLoopGroup-2-1] DEBUG io.netty.handler.ssl.SslHandler - [id: 0x318e5438, /xx.xx.xx.xxx:53458 => gateway.sandbox.push.apple.com/17.172.232.18:2195] HANDSHAKEN: SSL_RSA_WITH_AES_128_CBC_SHA
2015-02-18 09:18:30,232 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-1 sending known-bad notification to shut down.
2015-02-18 09:18:30,293 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - APNs gateway rejected notification with sequence number 0 from ExamplePushManager-connection-1 (MISSING_TOKEN).
2015-02-18 09:18:30,294 [nioEventLoopGroup-2-1] ERROR com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-1 failed to find rejected notification with sequence number 0 (buffer has range 1 to 1); this may mean the sent notification buffer is too small. Please report this as a bug.
2015-02-18 09:18:30,295 [pool-1-thread-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-2 beginning connection process.
2015-02-18 09:18:30,356 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-2 connected; waiting for TLS handshake.
2015-02-18 09:18:30,527 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-2 successfully completed TLS handshake.
2015-02-18 09:18:30,527 [nioEventLoopGroup-2-1] DEBUG io.netty.handler.ssl.SslHandler - [id: 0x54384092, /xx.xx.xx.xxx:53459 => gateway.sandbox.push.apple.com/17.172.232.18:2195] HANDSHAKEN: SSL_RSA_WITH_AES_128_CBC_SHA
2015-02-18 09:18:30,530 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-2 sending known-bad notification to shut down.
2015-02-18 09:18:30,592 [nioEventLoopGroup-2-1] DEBUG com.relayrides.pushy.apns.ApnsConnection - APNs gateway rejected notification with sequence number 0 from ExamplePushManager-connection-2 (MISSING_TOKEN).
2015-02-18 09:18:30,593 [nioEventLoopGroup-2-1] ERROR com.relayrides.pushy.apns.ApnsConnection - ExamplePushManager-connection-2 failed to find rejected notification with sequence number 0 (buffer has range 1 to 1); this may mean the sent notification buffer is too small. Please report this as a bug.
....

Thank you

**Comment by U-238** [comment]
I am only using the sandbox at this stage. I'm not yet in production.

**Comment by U-238** [comment]
I have fixed this in pull request #152

**Comment by varigamy** [comment]
Production environments (built on PROD certificates) work properly, only sandbox ones cause troubles

**Comment by jchambers** [comment]
Thanks, @U-238, for the fix. Will add some tests and put out a release as soon as I can.

**Comment by jchambers** [comment]
Added tests in d22a79b. Will put out a release tomorrow (I don't have my signing keys handy right now).

**Comment by jchambers** [comment]
In case you folks aren't on the mailing list, we've released Pushy v0.4.2 which includes a fix for this issue.
